### PR TITLE
[JSC] Add op_call_ignore_result

### PR DESCRIPTION
**Pull request by Constellation**
#### 61754abb0d67322e27e108d0d6fde7cf1a6ebc60
<pre>
[JSC] Add op_call_ignore_result
<a href="https://bugs.webkit.org/show_bug.cgi?id=258338">https://bugs.webkit.org/show_bug.cgi?id=258338</a>
rdar://111082116

Reviewed by Mark Lam.

This patch newly adds op_call_ignore_result. The concept is very simple: this call is procedure invocation, and
not expecting result from the call. This results in two effects.

1. Removing ValueProfile, dst register. This reduces profiling collection heaviness. Removing many baseline JIT instructions.
2. Telling that the result is not used to DFG nodes even though MovHint removal is not done (in DFG, it is not done. In FTL, it is done).
   Later, this can be leveraged, for example, we can add ArraySplice intrinsic. And by using this information, we can skip creating result arrays.

By adding this new op_call_ignore_result, CLoop&apos;s llint_cloop_did_return_from_js_xxx exceeds the threshold, so we separate them out of OpcodeID&apos;s &lt;= 256
space to make it work (they do not need to be &lt;= 256 since it is only used for LR).

* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/bytecode/BytecodeOperandsForCheckpoint.h:
(JSC::destinationFor):
* Source/JavaScriptCore/bytecode/BytecodeUseDef.cpp:
(JSC::computeUsesForBytecodeIndexImpl):
(JSC::computeDefsForBytecodeIndexImpl):
* Source/JavaScriptCore/bytecode/CallLinkInfo.cpp:
(JSC::CallLinkInfo::callTypeFor):
* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::finishCreation):
* Source/JavaScriptCore/bytecode/Opcode.h:
* Source/JavaScriptCore/bytecode/OpcodeInlines.h:
(JSC::isOpcodeShape):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::generate):
(JSC::BytecodeGenerator::BytecodeGenerator):
(JSC::BytecodeGenerator::emitCallInTailPosition):
(JSC::BytecodeGenerator::emitCall):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::emitNode):
(JSC::BytecodeGenerator::emitNodeInTailPositionFromReturnNode):
(JSC::BytecodeGenerator::emitNodeInTailPositionFromExprStatementNode):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::ExprStatementNode::emitBytecode):
(JSC::ReturnNode::emitBytecode):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::getPredictionWithoutOSRExit):
(JSC::DFG::ByteCodeParser::parseBlock):
* Source/JavaScriptCore/dfg/DFGGraph.cpp:
(JSC::DFG::Graph::methodOfGettingAValueProfileFor):
* Source/JavaScriptCore/dfg/DFGOSRExitCompilerCommon.cpp:
(JSC::DFG::callerReturnPC):
(JSC::DFG::reifyInlinedCallFrames):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
(JSC::DFG::SpeculativeJIT::compileFunction):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::emitCall):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::lower):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/ftl/FTLSlowPathCall.h:
(JSC::FTL::callOperation):
* Source/JavaScriptCore/jit/CallFrameShuffler.cpp:
(JSC::CallFrameShuffler::prepareAny):
* Source/JavaScriptCore/jit/JIT.cpp:
(JSC::JIT::privateCompileMainPass):
(JSC::JIT::privateCompileSlowCases):
* Source/JavaScriptCore/jit/JIT.h:
* Source/JavaScriptCore/jit/JITCall.cpp:
(JSC::JIT::compileSetupFrame):
(JSC::JIT::compileOpCall):
(JSC::JIT::emit_op_call_ignore_result):
(JSC::JIT::emitSlow_op_call_ignore_result):
* Source/JavaScriptCore/llint/LLIntThunks.cpp:
(JSC::LLInt::returnLocationThunk):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/FileBasedFuzzerAgent.cpp:
(JSC::FileBasedFuzzerAgent::getPredictionInternal):
* Source/JavaScriptCore/runtime/Gate.h:
* Source/JavaScriptCore/runtime/PredictionFileCreatingFuzzerAgent.cpp:
(JSC::PredictionFileCreatingFuzzerAgent::getPredictionInternal):

Canonical link: <a href="https://commits.webkit.org/266519@main">https://commits.webkit.org/266519@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/7afe070e9c083e49cc7d945e9932dbb254392edd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14032 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14347 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14682 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15770 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/13317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16856 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14431 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/15993 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14202 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14790 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11891 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16480 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/12076 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12652 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/19687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/11993 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13152 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12816 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16031 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/13302 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13366 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11230 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/14074 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12641 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3647 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16975 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/14463 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1656 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13209 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3461 "Passed tests") |

### CODE REVIEW

File: Source/JavaScriptCore/interpreter/CallFrame.h [hunk 1]
```diff
@@ -219,6 +221,16 @@ using JSInstruction = BaseInstruction<JSOpcodeTraits>;
         CallSiteIndex callSiteIndex() const;
         CallSiteIndex unsafeCallSiteIndex() const;
 
+        // This flag can be only used from host functions.
+        bool ignoreResult() const
```
Comment (Jarred-Sumner) [inline Source/JavaScriptCore/interpreter/CallFrame.h]: exciting

Comment (Constellation) [reply Source/JavaScriptCore/interpreter/CallFrame.h]: We first defer this thing to ensure that it is not perf regression. For now, I'll just add call_ignore_result, and then, leveraging this information in DFG or upper.

File: Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp [hunk 2]
```diff
@@ -337,6 +337,12 @@ ParserError BytecodeGenerator::generate(unsigned& size)
     RELEASE_ASSERT(m_codeBlock->numCalleeLocals() < static_cast<unsigned>(FirstConstantRegisterIndex));
     size = instructions().size();
     m_codeBlock->finalize(m_writer.finalize());
+
+    // We limit total bytecode sequence size in int32_t so that any int32_t jump offset can cover.
```
Comment (MenloDorian) [inline Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp]: /size in int32_t/size to int32_t/
/so that any int32_t jump offset can cover/so that we can use int32_t jump offsets/

File: Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp [hunk 3]
```diff
@@ -953,6 +953,14 @@ class ByteCodeParser {
...
             if (UNLIKELY(fuzzerAgent))
                 return fuzzerAgent->getPrediction(codeBlock, codeOrigin, prediction) & SpecBytecodeTop;
+
+            // If this instruction is derived from op_call_ignore_result, then we do not need to care about the result's prediction.
+            // Let's just return SpecFullTop to avoid SpecNone related ForceOSRExit.
+            auto instruction = codeBlock->instructions().at(codeOrigin.bytecodeIndex().offset());
+            OpcodeID opcodeID = instruction->opcodeID();
+            if (opcodeID == op_call_ignore_result)
+                return SpecFullTop;
+
```
Comment (MenloDorian) [inline Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp]: Should we do this check act the top instead?  If this is a `op_call_ignore_result`, then there's no need to compute `prediction` nor generate a fuzzer prediction, right?  It should always be `SpecFullTop` unconditionally.

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp [hunk 4]
```diff
@@ -638,6 +638,8 @@ void SpeculativeJIT::emitCall(Node* node)
     bool isDirect = false;
     switch (node->op()) {
     case DFG::Call:
+        callType = CallLinkInfo::Call;
+        break;
```
Comment (MenloDorian) [inline Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:642]: Why is this needed?

File: Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp [hunk 5]
```diff
@@ -406,6 +413,7 @@ BytecodeGenerator::BytecodeGenerator(VM& vm, FunctionNode* functionNode, Unlinke
     //
     // Note that we intentionally enable tail call for naked constructors since it does not have special code for "return".
     , m_inTailPosition(Options::useTailCalls() && !isConstructor() && constructorKind() == ConstructorKind::None && functionNode->isStrictMode())
+    , m_inCallIgnoreResultPosition(true)
```
Comment (MenloDorian) [inline Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp]: Why do you default `m_inCallIgnoreResultPosition` to true here?  Why not only `SetForScope` it to true in `emitNodeInTailPositionFromExprStatementNode`?  

I'm trying to understand why you went with the model where everything ignores results by default except a few places which opts out, vs a more conservative approach where everything does not ignore results (as before) except for the place where we know that it is not needed?  What insight am I missing that tells you that it's safe and correct to go with the opt out approach?

Comment (Constellation) [reply Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp]: Because I follow to the existing `m_inTailPosition` mechanism, to avoid introducing different logic and bugs.

File: Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp (hunk 3 above)
Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp]: This is necessary to be here since it is possible that the following tail-call's caller is `op_call_ignore_result`. So we need to check this inside `getValueProfilePredictionFromForCodeBlockAndBytecodeOffset`.

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp (hunk 4 above)
Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:642]: Ah, we can remove it. It was added to make the upcoming change easier.

File: Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp (hunk 5 above)
Comment (MenloDorian) [reply Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp]: After discussing this with Yusuke offline, I came to understand that:
1. Yusuke's goal is not to eliminate all unused function call results.  He's only implementing a conservative set, which happens to be a majority of naturally occurring cases.
2. The conservative set is the set of all tail calls that are not in return statements.  By definition, the only user of the result of a tail call is the enclosing return statement.  If the tail call is not in a return statement, then its result is, by definition, unused.

With that understanding, I suggest the following:

Suggestion #1: Rename `m_inCallIgnoreResultPosition` to `m_allowCallIgnoreResultOptimization`.   I think `m_inCallIgnoreResultPosition` is misleading because there are other cases of statements that may be in a position where its call result can be ignored, but yet we don't set this flag ... which is what led to my original confusion.  In contrast, `m_allowCallIgnoreResultOptimization` only says whether we allow the optimization or not, and makes no statement about whether a call is in a position where its result can be ignored or not.  I think this expresses the intent more clearly.

For that matter, I've come to see that `m_inTailPosition` is also a misnomer.  It should actually be named `m_allowTailCallOptimization` instead, but that's not related to this patch.

Suggestion #2: add a comment above this line indicating something like this:
```
// Currently, we're only conservatively allowing CallIgnoreResult optimization on tail call results that are
// not in return statements. We're not attempting to eliminate all unused call results.
```

**Review by MenloDorian — APPROVED** [review]
r=me
